Support `Scratch The OAP Config Dump`.

wu-sheng opened the pull request:
<!--
    ⚠️ Please make sure to read this template first, pull requests that don't accord with this template
    maybe closed without notice.
    Texts surrounded by `<` and `>` are meant to be replaced by you, e.g. <framework name>, <issue number>.
    Put an `x` in the `[ ]` to mark the item as CHECKED. `[x]`
-->

<!-- ==== 🐛 Remove this line WHEN AND ONLY WHEN you're fixing a bug, follow the checklist 👇 ====
### Fix <bug description or the bug issue number or bug issue link>
- [ ] Add a unit test to verify that the fix works.
- [ ] Explain briefly why the bug exists and how to fix it.
     ==== 🐛 Remove this line WHEN AND ONLY WHEN you're fixing a bug, follow the checklist 👆 ==== -->

<!-- ==== 📈 Remove this line WHEN AND ONLY WHEN you're improving the performance, follow the checklist 👇 ====
### Improve the performance of <class or module or ...>
- [ ] Add a benchmark for the improvement, refer to [the existing ones](https://github.com/apache/skywalking/blob/master/apm-commons/apm-datacarrier/src/test/java/org/apache/skywalking/apm/commons/datacarrier/LinkedArrayBenchmark.java)
- [ ] The benchmark result.
```text
<Paste the benchmark results here>
```
- [ ] Links/URLs to the theory proof or discussion articles/blogs. <links/URLs here>
     ==== 📈 Remove this line WHEN AND ONLY WHEN you're improving the performance, follow the checklist 👆 ==== -->

<!-- ==== 🆕 Remove this line WHEN AND ONLY WHEN you're adding a new feature, follow the checklist 👇 ====
### <Feature description>
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.
- [ ] Update the documentation to include this new feature.
- [ ] Tests(including UT, IT, E2E) are added to verify the new feature.
- [ ] If it's UI related, attach the screenshots below.
     ==== 🆕 Remove this line WHEN AND ONLY WHEN you're adding a new feature, follow the checklist 👆 ==== -->

- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).

wu-sheng commented:
The meaningful config dump

```shell
> curl http://127.0.0.1:12800/debugging/config/dump
cluster.provider=standalone
core.provider=default
core.default.prepareThreads=2
core.default.restHost=0.0.0.0
core.default.searchableLogsTags=level,http.status_code
core.default.role=Mixed
core.default.persistentPeriod=25
core.default.syncPeriodHttpUriRecognitionPattern=10
core.default.restIdleTimeOut=30000
core.default.dataKeeperExecutePeriod=5
core.default.topNReportPeriod=10
core.default.gRPCSslTrustedCAPath=
core.default.downsampling=[Hour, Day]
core.default.serviceNameMaxLength=70
core.default.gRPCSslEnabled=false
core.default.restPort=12800
core.default.serviceCacheRefreshInterval=10
...
```

kezhenxu94 commented on oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModuleConfig.java:
```diff
@@ -224,6 +224,21 @@ public class CoreModuleConfig extends ModuleConfig {
+     * @since 9.7.0
+     */
+    @Getter
+    private String keywords4MaskingSecretsOfConfig = "user,password,token,accessKey,secretKey}";
```
```suggestion
    private String keywords4MaskingSecretsOfConfig = "user,password,token,accessKey,secretKey";
```

kezhenxu94 commented on oap-server/server-starter/src/main/resources/application.yml:
```diff
@@ -137,6 +137,10 @@ core:
     trainingPeriodHttpUriRecognitionPattern: ${SW_CORE_TRAINING_PERIOD_HTTP_URI_RECOGNITION_PATTERN:60}
     # The max number of HTTP URIs per service for further URI pattern recognition.
     maxHttpUrisNumberPerService: ${SW_CORE_MAX_HTTP_URIS_NUMBER_PER_SVR:3000}
+    # Include the list of keywords to filter configurations including secrets. Separate keywords by a comma.
+    keywords4MaskingSecretsOfConfig: ${SW_CORE_KEYWORDS_FOR_MASKING_SECRETS:user,password,token,accessKey,secretKey}
+    # Disable core HTTP:port/debugging/config/dump
+    disableConfigDump: ${SW_DISABLE_CONFIG_DUMP:false}
```
These are all about the new module, why should we add them in the core section?

wu-sheng replied:
Because ServerStatusService is a core module service.

kezhenxu94 replied:
> I am thinking of the other features, such as L1/L2 queue status dump, and cache dump.

I believe those should be done through self-observability metrics

kezhenxu94 commented on oap-server/server-starter/src/main/resources/application.yml:
```diff
@@ -535,6 +539,10 @@ health-checker:
     checkIntervalSeconds: ${SW_HEALTH_CHECKER_INTERVAL_SECONDS:5}
 
+debugging-query:
+  selector: ${SW_DEBUGGING_QUERY:default}
```
For such a plugin for debugging purpose, I would say users only want to enable it when they are debugging something, and should be turned off after debugging, I don't think there is any reason to turn it on by default

wu-sheng replied:
Changing the value means rebooting, I don't want that happen.

kezhenxu94 replied:
> Changing the value means rebooting, I don't want that happen.

I think it's a common sense that debugging means `do adjust codes or configs; reboot; until it works as expected`, expose these configs might be risky so if anyone missed a keyword (like https://github.com/apache/skywalking/pull/11579/files#r1402079849), they have to reboot their production env.

kezhenxu94 reviewed (changes requested):
Looks like the implementation doesn't take care the dynamic configurations at all, when a config key `K`'s default is `V` and is set to `V2` in dynamic configuration, the dump API always returns `V` (instead of `V2` as expected), which is misleading.

wu-sheng commented:
> Looks like the implementation doesn't take care the dynamic configurations at all, when a config key `K`'s default is `V` and is set to `V2` in dynamic configuration, the dump API always returns `V` (instead of `V2` as expected), which is misleading.

No, it would not. Because the dynamic configuration is never overriding the value of the config. There are two values in there. 
This is a honest dump, nothing more, nothing less.

wu-sheng commented:
And as I mentioned at https://github.com/apache/skywalking/pull/11579#discussion_r1402062433, there should be more debugging relative APIs exposed.

kezhenxu94 commented on oap-server/server-starter/src/main/resources/application.yml:
```diff
@@ -137,6 +137,10 @@ core:
     # The max number of HTTP URIs per service for further URI pattern recognition.
     maxHttpUrisNumberPerService: ${SW_CORE_MAX_HTTP_URIS_NUMBER_PER_SVR:3000}
+    # Include the list of keywords to filter configurations including secrets. Separate keywords by a comma.
+    keywords4MaskingSecretsOfConfig: ${SW_CORE_KEYWORDS_FOR_MASKING_SECRETS:user,password,token,accessKey,secretKey}
```
`authentication` is also a secret for some of the configs (for example, `receiver-sharing-server.default.authentication`)

wu-sheng commented:
I updated the codes
1. Moved the config `keywords4MaskingSecretsOfConfig` to the `debugging-query-plugin`.
2. Removed `SW_DISABLE_CONFIG_DUMP`, as the module itself could be disabled directly.
3. Added `authentication` as a keyword.
4. Docs updated accordingly.
